### PR TITLE
chore: re-re-implementation of soft deletion to catalog query/content…

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
… metadata relationships

## Description

Description of changes made

## Ticket Link

Link to the associated ticket
[Link title](https://openedx.atlassian.net/browse/ENT-XXXX)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        ),
+        migrations.RunSQL(
+            sql="""
+                /* there's some funky stuff with defaults and the timestamped model */
+                INSERT INTO catalog_contentmetadatatoqueries (catalog_query_id, content_metadata_id, created, modified)
+                SELECT catalogquery_id, contentmetadata_id, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP FROM catalog_contentmetadata_catalog_queries;
+            """,
+            # Noop the reverse sql as this table is large enough to cause issues with a large insert and preserving
+            # these particular records on revert is not of high priority
+            reverse_sql=migrations.RunSQL.noop
```
Comment (alex-sheehan-edx) [inline enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py:42]: removed backwards migration on the data from the table as it was causing too many issues on prod

File: enterprise_catalog/apps/catalog/models.py [hunk 2]
```diff
@@ -117,6 +119,14 @@ def __str__(self):
             )
         )
 
+    @property
+    def content_metadata(self):
+        # Join catalog queries on content metadata through the through table where deleted at is null
+        return ContentMetadata.objects.filter(
```
Comment (alex-sheehan-edx) [inline enterprise_catalog/apps/catalog/models.py:124]: changed this from a scan to a join

File: enterprise_catalog/apps/catalog/models.py [hunk 3]
```diff
@@ -438,6 +448,14 @@ def recently_modified_records(cls, time_delta):
             modified__range=(range_start, range_end),
         )
 
+    @property
+    def catalog_queries(self):
+        # Join content metadata on catalog queries through the through table where deleted at is null
+        return CatalogQuery.objects.filter(
```
Comment (alex-sheehan-edx) [inline enterprise_catalog/apps/catalog/models.py:453]: same scan -> join logic to improve efficiency of queries

File: enterprise_catalog/apps/catalog/models.py [hunk 4]
```diff
@@ -449,6 +467,87 @@ def __str__(self):
...
+    )
+    content_metadata = models.ForeignKey(
+        ContentMetadata,
+        on_delete=models.deletion.DO_NOTHING,
+    )
+
+    class Meta:
+        app_label = 'catalog'
+        unique_together = ('content_metadata_id', 'catalog_query_id', 'deleted_at')
+        index_together = (
```
Comment (alex-sheehan-edx) [inline enterprise_catalog/apps/catalog/models.py:544]: added two additional indices to aid in querying efficiency

File: enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('created', model_utils.fields.AutoCreatedField(default=django.utils.timezone.now, editable=False, verbose_name='created')),
+                ('modified', model_utils.fields.AutoLastModifiedField(default=django.utils.timezone.now, editable=False, verbose_name='modified')),
+                ('deleted_at', models.DateTimeField(blank=True, null=True)),
+                ('catalog_query', models.ForeignKey(on_delete=django.db.models.deletion.DO_NOTHING, to='catalog.catalogquery')),
+                ('content_metadata', models.ForeignKey(on_delete=django.db.models.deletion.DO_NOTHING, to='catalog.contentmetadata')),
+            ],
+            options={
+                'unique_together': {('content_metadata_id', 'catalog_query_id', 'deleted_at')},
+                'index_together': {('catalog_query_id', 'deleted_at'), ('content_metadata_id', 'deleted_at')},
```
Comment (johnnagro) [inline enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py:26]: this should cover the join queries we were seeing "using where" the other night

File: enterprise_catalog/apps/catalog/models.py (hunk 2 above)
Comment (johnnagro) [reply enterprise_catalog/apps/catalog/models.py:124]: the old way was resulting in big IN queries passing ~2700+ ids. not a good utilization of indexing, etc. this way is an actual join statement under the covers.

File: enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py (hunk 5 above)
Comment (alex-sheehan-edx) [reply enterprise_catalog/apps/catalog/migrations/0028_contentmetadatatoqueries.py:26]: 🤞

**Review by johnnagro — APPROVED** [review]

**Pull request merged**